Automattic/woocommerce-payments
# Enable deferred intent creation when initialization process encounters cache unavailability

Fixes https://github.com/Automattic/woocommerce-payments/issues/7482

#### Changes proposed in this Pull Request

When checking the state of the feature flag for the deferred intent creation UPE, sometimes the cache isn't available which makes the [`WC_Payments_Features` return false](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-features.php#L81). If with this setup, a store has legacy UPE enabled, [legacy UPE takes over the initialization](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L548) (instead of [initializing deferred UPE](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments.php#L527) as one would expect), eventually leading to unpredictable state of the registered gateways. This is what the author of https://github.com/Automattic/woocommerce-payments/issues/7482 faced. 

The main change in this PR is to enable deferred UPE when the cache is unavailable for some reason. Cache being unavailable is a rare/edge-case scenario, and the changes in this PR are not negatively impacting the scenario when we disable the dUPE feature flag for a merchant. The flag state is correctly propagated to the plugin if we do so.

While the unit test suite is initialized, it checks UPE flags. Previously, while checking the flag, all the UPEs were disabled, and thus, the legacy gateway was initialized for tests. Now, when we run dUPE in case of a non-available cache, tests start to run deferred intent creation UPE, which led to multiple failures that this PR fixes. I encourage to review the tests in more detail. Some tests were removed, some were replaced, and some context is left in the comments. 









#### Testing instructions

On `develop`
1. Ensure that the server-side flag for dUPE on your local store is enabled
2. Enable client-side FF for legacy UPE
3. In WCPay Dev, refresh the cache and confirm you see a fatal error like described https://github.com/Automattic/woocommerce-payments/issues/7482#issuecomment-1784989547
4. Checkout `fix/more-use-cases-for-enabled-dupe` 
5. Repeat steps 1-3 and confirm that you don't see the error from the third step anymore
6. Perform a sanity check by purchasing a product and confirm that deferred intent creation UPE is being used (e.g. by checking the `gateway_type` field in the Metadata of payment details on the Stripe dashboard)
7. Disable dUPE FF in the server's database and confirm that legacy UPE is used if legacy UPE FF is on, and legacy card is used when no UPE FF is enabled)
8. Run `npm run test:php` on your local to ensure that the changes in `bootstrap.php` don't impact the setup negatively. 
-------------------


**Post merge**
